### PR TITLE
Add state to compiler visitors, add optimizations for memory file system and parsing

### PR DESCRIPTION
**Pull request by sebmck**
## Summary

See each individual commit title for specific summary of changes.

## Test Plan

CI. Numerous manual local tests.

### CODE REVIEW

File: internal/compiler/lint/rules/jsx-a11y/mouseEventsHaveKeyEvents.test.rjson [hunk 1]
```diff
@@ -6,8 +6,8 @@ invalid: [
 	"<div onMouseOut={ () => void 0 } >{props}</div>"
 ]
 valid: [
-	"<div onMouseOver={ () => void 0 } onFocus={() => void 0} />"
-	"<div onMouseOut={ () => void 0 } onBlur={() => void 0} />"
-	"<div onMouseOver={ () => void 0 }  onFocus={() => void 0} >{props}</div>"
-	"<div onMouseOut={ () => void 0 }  onBlur={() => void 0} >{props}</div>"
+	"<div onMouseOver={ () => void 0 } onFocus={VoidCallback 0} />"
```
Comment (ematipico) [inline internal/compiler/lint/rules/jsx-a11y/mouseEventsHaveKeyEvents.test.rjson]: @sebmck  Is this valid? `VoidCallback` is a type and the snapshots removed some valid example

Comment (sebmck) [reply internal/compiler/lint/rules/jsx-a11y/mouseEventsHaveKeyEvents.test.rjson]: It's not, I ended up fixing it in a different PR.

File: internal/compiler/transforms/helpers/index.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
+/**
```
Comment (ematipico) [inline internal/compiler/transforms/helpers/index.ts:1]: The comment header could be removed

Comment (sebmck) [reply internal/compiler/transforms/helpers/index.ts:1]: It shouldn't. It's a rename of existing file, any derivative files of existing source code need to have the copyright header retained.

**Pull request merged**